envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/listeners/filters/RBACFilterFactory.kt

```diff
@@ -0,0 +1,102 @@
+) {
+    companion object {
+        private val logger by logger()
+    }
+
+    fun getRules(serviceName: String, incomingPermissions: Incoming): RBAC {
+        val clientToPolicy = mutableMapOf<String, Policy>()
+        incomingPermissions.endpoints.forEach { incomingEndpoint ->
+            if (incomingEndpoint.clients.isEmpty()) {
+                logger.error("An incoming endpoint definition for $serviceName does not have any clients defined." +
+                        "It means that noone will be able to contact that endpoint.")
+                return@forEach
+            }
+
+            val policy = Policy.newBuilder()
+            val clientName = incomingEndpoint.clients.joinToString(",")
+
+            val pathPermission = Permission.newBuilder().setHeader(getPathMatcher(incomingEndpoint))
+            val methodPermissions = Permission.newBuilder()
+            val combinedPermissions = Permission.newBuilder()
+
+            val principals = incomingEndpoint.clients.map(this::mapClientToPrincipal)
+            policy.addAllPrincipals(principals)
+
+            if (incomingEndpoint.methods.isNotEmpty()) {
+                val methodPermissionSet = Permission.Set.newBuilder()
+                val methodPermissionsList = incomingEndpoint.methods.map(this::mapMethodToHeaderMatcher)
+                methodPermissionSet.addAllRules(methodPermissionsList)
+                methodPermissions.setOrRules(methodPermissionSet.build())
+                combinedPermissions.setAndRules(
+                        Permission.Set.newBuilder().addAllRules(listOf(
+                                pathPermission.build(),
+                                methodPermissions.build()
+                        ))
+                )
+            } else {
+                combinedPermissions.setAndRules(
+                        Permission.Set.newBuilder().addAllRules(listOf(pathPermission.build()))
+                )
+            }
+
+            policy.addPermissions(combinedPermissions)
+            clientToPolicy[clientName] = policy.build()
+        }
+
+        val rbac = RBAC.newBuilder()
+                .setAction(RBAC.Action.ALLOW)
+                .putAllPolicies(clientToPolicy)
+                .build()
+        return rbac
+    }
+
+    private fun mapMethodToHeaderMatcher(method: String): Permission {
+        val methodMatch = HeaderMatcher.newBuilder().setName(":method").setExactMatch(method).build()
+        return Permission.newBuilder().setHeader(methodMatch).build()
+    }
+
+    private fun mapClientToPrincipal(client: String): Principal {
+        val clientMatch = HeaderMatcher.newBuilder()
+                .setName(properties.clientIdentityHeader).setExactMatch(client).build()
+        return Principal.newBuilder().setHeader(clientMatch).build()
+    }
+
+    private fun getPathMatcher(incomingEndpoint: IncomingEndpoint): HeaderMatcher {
+        return when (incomingEndpoint.pathMatchingType) {
```

Comment:
What about regex_match? There are services that use paths like `/resource/{id}/action` and Envoy supports it.

Reply:
I think the initial implementation was missing it as well. See: https://github.com/allegro/envoy-control/blob/797741e660bcced4303c608d53e2124cbd49c84a/envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt#L236 

@MarcinFalkowski - do you recommend using PATH_PREFIX for that or do we have some special case here?